### ISSUE TITLE
No display when connecting a Meizu phone

### ISSUE THREAD

**Issue by FredWe** [Genymobile/scrcpy#240]
same issue as comment in #151 

Scrcpy: v1.3
Device: Meizu Pro 6 plus
Android version: 7.0.0
ROM: Flyme 7.8.7.24 beta
Host: Ubuntu 18.04

screenrecording by adb works well as following:
```
adb shell screenrecord /sdcard/test.mp4
adb pull /sdcard/test.mp4
```

But issue appeared when executing
```
./run x        
x/server/scrcpy-server.jar: 1 file pushed. 3.0 MB/s (18570 bytes in 0.006s)
DEBUG: Starting decoder thread
DEBUG: Starting controller thread
INFO: OpenGL shaders: ENABLED
INFO: Created renderer: opengl
INFO: Initial texture: 1440x2560
```

and hold on here... no response, no display window (as i expected)

and i got some logcat related to maybe a ROM bug:
```
08-23 14:01:00.196  2650  3656 E OMXNodeInstance: getConfig(a5a00bd:Exynos.AVC.Encoder, ConfigAndroidIntraRefresh(0x6f60000a)) ERROR: UnsupportedIndex(0x8000101a)
08-23 14:01:00.200 18461 18461 E Surface : Could not create instance of Surface from IGraphicBufferProducer.
08-23 14:01:00.201 18461 18461 E Surface : java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
08-23 14:01:00.201 18461 18461 E Surface : 	at android.os.Handler.<init>(Handler.java:200)
08-23 14:01:00.201 18461 18461 E Surface : 	at android.os.Handler.<init>(Handler.java:114)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.meizu.nightmode.MzNightModeManager.<init>(MzNightModeManager.java:148)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.meizu.nightmode.MzNightModeManager.getDefault(MzNightModeManager.java:262)
08-23 14:01:00.201 18461 18461 E Surface : 	at android.graphics.Canvas.<init>(Canvas.java:2606)
08-23 14:01:00.201 18461 18461 E Surface : 	at android.view.Surface$CompatibleCanvas.<init>(Surface.java:595)
08-23 14:01:00.201 18461 18461 E Surface : 	at android.view.Surface.<init>(Surface.java:93)
08-23 14:01:00.201 18461 18461 E Surface : 	at android.media.MediaCodec.createInputSurface(Native Method)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.genymobile.scrcpy.ScreenEncoder.streamScreen(ScreenEncoder.java:63)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.genymobile.scrcpy.Server.scrcpy(Server.java:24)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.genymobile.scrcpy.Server.main(Server.java:102)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
08-23 14:01:00.201 18461 18461 E Surface : 	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:277)
```

Just for your reference, or for other who met the same issue.

**Comment by zmjios** [comment]
i am also meet this question,can anybody reslove this?

**Comment by rom1v** [comment]
cf https://github.com/Genymobile/scrcpy/pull/434

**Comment by rom1v** [comment]
It has been fixed some time ago.

**Comment by Stanislavoloshchuk** [comment]
I download archive and tried instructions(as far as I understood it) but gives following 

```
stanislav@stanislav:~$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1.13
```
Then I tried to extract and add the path of extracted files:
```
ERROR: Neither directory contains a build file meson.build.

stanislav@stanislav:~$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1

ERROR: Neither directory contains a build file meson.build.

stanislav@stanislav:~$ ninja -Cx
ninja: Entering directory `x'
ninja: error: loading 'build.ninja': No such file or directory

stanislav@stanislav:~$ sudo ninja -Cx install
[sudo] password for stanislav: 
ninja: Entering directory `x'
ninja: error: loading 'build.ninja': No such file or directory
```
Seems I am doing something wrong...

**Comment by rom1v** [comment]
> I download archive and tried instructions(as far as I understood it) but gives following

You need to clone the repo beforehand:

```
git clone https://github.com/Genymobile/scrcpy
cd scrcpy
meson ...
```

**Comment by Stanislavoloshchuk** [comment]
Hello! So I have done all those steps, and what we have here:
```stanislav@stanislav:~/scrcpy$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1
The Meson build system
Version: 0.53.2
Source dir: /home/stanislav/scrcpy
Build dir: /home/stanislav/scrcpy/x
Build type: native build
Project name: scrcpy
Project version: 1.14
C compiler for the host machine: cc (gcc 9.3.0 "cc (Ubuntu 9.3.0-10ubuntu2) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency libavformat found: YES 58.29.100
Run-time dependency libavcodec found: YES 58.54.100
Run-time dependency libavutil found: YES 56.31.100
Run-time dependency sdl2 found: YES 2.0.10
Configuring config.h using configuration
server/meson.build:19: WARNING: Custom target input '/home/stanislav/Downloads/scrcpy-server-v1' can't be converted to File object(s).
This will become a hard error in the future.
Build targets in project: 3

Found ninja-1.10.0 at /usr/bin/ninja
stanislav@stanislav:~/scrcpy$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1.13
Directory already configured.

Just run your build command (e.g. ninja) and Meson will regenerate as necessary.
If ninja fails, run "ninja reconfigure" or "meson --reconfigure"
to force Meson to regenerate.

If build failures persist, run "meson setup --wipe" to rebuild from scratch
using the same options as passed when configuring the build.
To change option values, run "meson configure" instead.
stanislav@stanislav-UX32VD:~/scrcpy$ ninja -Cx
ninja: Entering directory `x'
[20/26] Generating scrcpy-server-prebuilt with a custom command.
FAILED: server/scrcpy-server 
cp /home/stanislav/Downloads/scrcpy-server-v1 server/scrcpy-server
cp: -r not specified; omitting directory '/home/stanislav/Downloads/scrcpy-server-v1'
[25/26] Compiling C object 'app/a172ced@@scrcpy@exe/src_tiny_xpm.c.o'.
ninja: build stopped: subcommand failed.
stanislav@stanislav:~/scrcpy$ sudo ninja -Cx install
ninja: Entering directory `x'
[1/3] Generating scrcpy-server-prebuilt with a custom command.
FAILED: server/scrcpy-server 
cp /home/stanislav/Downloads/scrcpy-server-v1 server/scrcpy-server
cp: -r not specified; omitting directory '/home/stanislav/Downloads/scrcpy-server-v1'
[2/3] Linking target app/scrcpy.
ninja: build stopped: subcommand failed.
```

**Comment by rom1v** [comment]
> Directory already configured.

`rm -rf x` and retry.

**Comment by Stanislavoloshchuk** [comment]
So, I just entered ```rm -rf x``` then call ```scrcpy``` and nothing except ```INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
/usr/local/share/scrcpy/scrcpy-serve...ed. 5.2 MB/s (33142 bytes in 0.006s)
[server] INFO: Device: Meizu M6 Note (Android 7.1.2)
INFO: OpenGL shaders: ENABLED
INFO: Created renderer: opengl
INFO: Renderer: opengl
INFO: OpenGL version: 4.4.0 NVIDIA 340.108
INFO: Trilinear filtering enabled
INFO: Initial texture: 1080x1920
``` happens =( sorry if I am mising some simple points I am noob in this kind of stuff.

**Comment by rom1v** [comment]
> So, I just entered `rm -rf x`

Then you must re-execute the meson/ninja commands.

Then, to run your build without installing, just execute `./run x`.

**Comment by Stanislavoloshchuk** [comment]
As I get:
```
stanislav@stanislav:~/scrcpy$ rm -rf x
```
then
```
stanislav@stanislav:~/scrcpy$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1
The Meson build system
Version: 0.53.2
Source dir: /home/stanislav/scrcpy
Build dir: /home/stanislav/scrcpy/x
Build type: native build
Project name: scrcpy
Project version: 1.14
C compiler for the host machine: cc (gcc 9.3.0 "cc (Ubuntu 9.3.0-10ubuntu2) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency libavformat found: YES 58.29.100
Run-time dependency libavcodec found: YES 58.54.100
Run-time dependency libavutil found: YES 56.31.100
Run-time dependency sdl2 found: YES 2.0.10
Configuring config.h using configuration
server/meson.build:19: WARNING: Custom target input '/home/stanislav/Downloads/scrcpy-server-v1' can't be converted to File object(s).
This will become a hard error in the future.
Build targets in project: 3

Found ninja-1.10.0 at /usr/bin/ninja
```
```
stanislav@stanislav:~/scrcpy$ ninja -Cx
ninja: Entering directory `x'
[20/26] Generating scrcpy-server-prebuilt with a custom command.
FAILED: server/scrcpy-server 
cp /home/stanislav/Downloads/scrcpy-server-v1 server/scrcpy-server
cp: -r not specified; omitting directory '/home/stanislav/Downloads/scrcpy-server-v1'
[25/26] Compiling C object 'app/a172ced@@scrcpy@exe/src_tiny_xpm.c.o'.
ninja: build stopped: subcommand failed.
```
```
stanislav@stanislav~/scrcpy$ sudo ninja -Cx install
ninja: Entering directory `x'
[1/3] Generating scrcpy-server-prebuilt with a custom command.
FAILED: server/scrcpy-server 
cp /home/stanislav/Downloads/scrcpy-server-v1 server/scrcpy-server
cp: -r not specified; omitting directory '/home/stanislav/Downloads/scrcpy-server-v1'
[2/3] Linking target app/scrcpy.
ninja: build stopped: subcommand failed.
```
```
stanislav@stanislav-UX32VD:~/scrcpy$ ./run x
INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
stat: No such file or directory
ERROR: 'x/server/scrcpy-server' does not exist or is not a regular file
```
is this correct actions?

**Comment by rom1v** [comment]
> -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1

The file is `scrcpy-server-v1.13`, not `scrcpy-server-v1`.

**Comment by Stanislavoloshchuk** [comment]
Alright, now seems to be all correct, let`s try it =)

```
stanislav@stanislav:~/scrcpy$ meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server-v1.13
The Meson build system
Version: 0.53.2
Source dir: /home/stanislav/scrcpy
Build dir: /home/stanislav/scrcpy/x
Build type: native build
Project name: scrcpy
Project version: 1.14
C compiler for the host machine: cc (gcc 9.3.0 "cc (Ubuntu 9.3.0-10ubuntu2) 9.3.0")
C linker for the host machine: cc ld.bfd 2.34
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency libavformat found: YES 58.29.100
Run-time dependency libavcodec found: YES 58.54.100
Run-time dependency libavutil found: YES 56.31.100
Run-time dependency sdl2 found: YES 2.0.10
Configuring config.h using configuration
Build targets in project: 3

Found ninja-1.10.0 at /usr/bin/ninja
```
```
stanislav@stanislav:~/scrcpy$ ninja -Cx
ninja: Entering directory `x'
[26/26] Linking target app/scrcpy.
```
```
stanislav@stanislav:~/scrcpy$ sudo ninja -Cx install
ninja: Entering directory `x'
[0/1] Installing files.
Installing app/scrcpy to /usr/local/bin
Stripping target 'app/scrcpy' using strip.
Installing server/scrcpy-server to /usr/local/share/scrcpy
Installing /home/stanislav/scrcpy/app/scrcpy.1 to /usr/local/share/man/man1
```
```
stanislav@stanislav:~/scrcpy$ ./run x
INFO: scrcpy 1.14 <https://github.com/Genymobile/scrcpy>
x/server/scrcpy-server: 1 file pushed. 0.7 MB/s (27694 bytes in 0.040s)
[server] ERROR: Exception on thread Thread[main,5,main]
java.lang.IllegalArgumentException: The server version (1.13) does not match the client (1.14)
	at com.genymobile.scrcpy.Server.createOptions(Server.java:78)
	at com.genymobile.scrcpy.Server.main(Server.java:176)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:281)

```

**Comment by Stanislavoloshchuk** [comment]
So, nothing =(
[log.txt](https://github.com/Genymobile/scrcpy/files/4767329/log.txt)

**Comment by rom1v** [comment]
> The server version (1.13) does not match the client (1.14)

You're compiling the current `master` of the client. You must compile v1.13.

```bash
git checkout v1.13
ninja -Cx
./run x
```

**Comment by Stanislavoloshchuk** [comment]
Genius! =) Many-many thanks! I owe you! =)

**Comment by rom1v** [comment]
Does v1.13 work better than v1.14?

**Comment by Stanislavoloshchuk** [comment]
Actually, seems the only 1.13 works, as I understood it was last recommended and it works.

**Comment by rom1v** [comment]
OK, then there is a regression in v1.14 (the latest stable), that's bad.

I need your help to fix now :)

Could you confirm that if you replace `x/server/scrcpy-server` (in your scrcpy directory) by this one: https://github.com/Genymobile/scrcpy/issues/365#issuecomment-621970307
then it fails again.

**Comment by Stanislavoloshchuk** [comment]
Sure I will try my best, I already downloaded from #365 (comment) and now could you advice how to find this ```x/server/scrcpy-server``` directory to replace it ?

**Comment by Stanislavoloshchuk** [comment]
Oh, found! =)

**Comment by Stanislavoloshchuk** [comment]
So, yeah I am confirming after replacement it is not working =(
But before replacement I copied original folder and now trying to replace it back-again and it doesn`t work anymore.

**Comment by rom1v** [comment]
OK so the culprit is 62c0c1321feec0a673329cb6771efeefb6dc9657 :/

**Comment by rom1v** [comment]
Too bad that your logs are not complete.

Could you try to clear the logs again:
```bash
adb logcat -c
./run x  # the version which fails
adb logcat -d > log.txt
```

**Comment by Stanislavoloshchuk** [comment]
After all these steps log is empty now

**Comment by rom1v** [comment]
Your logcat may have crashed. Could you reboot the device and try again, please?

**Comment by Stanislavoloshchuk** [comment]
okay lets try this one, but I am not exactly sure that this is one which was crashed because none is working now...
[log.txt](https://github.com/Genymobile/scrcpy/files/4767552/log.txt)

**Comment by rom1v** [comment]
OK so your logs seems limited to 1100 lines, then nothing :/ Very annoying.
Is it the same with just `adb logcat` (without `-d …`)?

> none is working now

You probably overwrote server v1.13 :)

Just to be sure, could you please try this one (it's v1.13 + the culprit commit in v1.14, without anything else):

 - [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/240/1/scrcpy-server)  
  _SHA256: 24b5ddd38641541122a4b7437fb04f3948d013765a14edf0aad8a74aeac9c194_

**Comment by Stanislavoloshchuk** [comment]
Here is ```log``` and ```log-d``` txts
[log.txt](https://github.com/Genymobile/scrcpy/files/4767583/log.txt)
[log-d.txt](https://github.com/Genymobile/scrcpy/files/4767584/log-d.txt)

**Comment by Stanislavoloshchuk** [comment]
could you please advice is I should do anything with this: ```SHA256: 24b5ddd38641541122a4b7437fb04f3948d013765a14edf0aad8a74aeac9c194```?

**Comment by rom1v** [comment]
It's the [checksum](https://en.wikipedia.org/wiki/Checksum), so that you can check that the file content is exactly the expected one.

Once you downloaded the file, execute:
```
sha256sum scrcpy-server
```

The sum should be the same.

It avoids corruption, and it mitigate risks of malicious downloads (one would have to modify both the file on my server and the checksum on github).

**Comment by Stanislavoloshchuk** [comment]
Here is checked sum:
```sha256sum scrcpy-server1
24b5ddd38641541122a4b7437fb04f3948d013765a14edf0aad8a74aeac9c194  scrcpy-server1```
seems to be ok

So to make sure now I have to:
```rm -rf x```
```meson x --buildtype release --strip -Db_lto=true     -Dprebuilt_server=/home/stanislav/Downloads/scrcpy-server1 ``` # lately downloaded from your message 10 minutes ago
```ninja -Cx```
```sudo ninja -Cx install```
```./run x```

Am I right?

**Comment by rom1v** [comment]
Yes (but no need to `sudo ninja -Cx install` if you just run it with `./run x`)

**Comment by Stanislavoloshchuk** [comment]
So:
```
INFO: scrcpy 1.13 <https://github.com/Genymobile/scrcpy>
x/server/scrcpy-server: 1 file pushed. 3.4 MB/s (38916 bytes in 0.011s)
[server] INFO: Device: Meizu M6 Note (Android 7.1.2)
INFO: Renderer: opengl
INFO: OpenGL version: 4.4.0 NVIDIA 340.108
INFO: Trilinear filtering enabled
INFO: Initial texture: 1080x1920
```
and thats it nothing else happens

**Comment by rom1v** [comment]
OK.

@FredWe (240) and @olbb (#434), could you test v1.14 on your Meizu, and (if it fails for you) post your logs. Thank you for your help :)

**Comment by Stanislavoloshchuk** [comment]
Could we please continue tomorrow? I just falling asleep now had a rush day. I will reply asap.

**Comment by Stanislavoloshchuk** [comment]
Hello. So shall I do anything or this is questions only for @FredWe and @olbb ?

**Comment by rom1v** [comment]
The request is for anyone having a Meizu having the problem, with a working `adb logcat`. :wink:

**Comment by Stanislavoloshchuk** [comment]
Ah, alright! Then thank you ;)

**Comment by rom1v** [comment]
@Stanislavoloshchuk Just in case, could you try this (replace in v1.14), please?

 - [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/240/2/scrcpy-server)
  _SHA256: eed7226eb179676c84812a8793cbb4946c5ff0a0f00ce7f031980de73a1dae53_

<details>
<summary>diff</summary>

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index d722388c..0da689d8 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -53,11 +53,11 @@ public class ScreenEncoder implements Device.RotationListener {
 
         try {
             internalStreamScreen(device, fd);
-        } catch (NullPointerException e) {
+        } catch (Throwable e) {
             // Retry with workarounds enabled:
             // <https://github.com/Genymobile/scrcpy/issues/365>
             // <https://github.com/Genymobile/scrcpy/issues/940>
-            Ln.d("Applying workarounds to avoid NullPointerException");
+            Ln.d("Applying workarounds to avoid exceptions");
             Workarounds.fillAppInfo();
             internalStreamScreen(device, fd);
         }
```
</details>

**Comment by olbb** [comment]
> Hello. So shall I do anything or this is questions only for @FredWe and @olbb ?

I'm sorry, but mine Meizu phone is damaged.

**Comment by FredWe** [comment]
> OK.
> 
> @FredWe and @olbb (#434), could you test v1.14 on your Meizu, and (if it fails for you) post your logs. Thank you for your help :)

Sorry for late response, i will test it after several hours.

**Comment by rom1v** [comment]
I plan to release a new version soon (v1.15), with or without this issue. If someone having this problem could find where (on which line) the server is blocked in v1.14, it would help.

**Comment by rom1v** [comment]
I added some logs.

<details>
<summary>diff</summary>

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index d722388c..a7855f3b 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -49,17 +49,27 @@ public class ScreenEncoder implements Device.RotationListener {
     }
 
     public void streamScreen(Device device, FileDescriptor fd) throws IOException {
+        Ln.i("### 1");
         Workarounds.prepareMainLooper();
+        Ln.i("### 2");
 
         try {
             internalStreamScreen(device, fd);
+            Ln.i("### 3");
         } catch (NullPointerException e) {
+            Ln.i("### 4");
             // Retry with workarounds enabled:
             // <https://github.com/Genymobile/scrcpy/issues/365>
             // <https://github.com/Genymobile/scrcpy/issues/940>
             Ln.d("Applying workarounds to avoid NullPointerException");
+            Ln.i("### 5");
             Workarounds.fillAppInfo();
+            Ln.i("### 6");
             internalStreamScreen(device, fd);
+            Ln.i("### 7");
+        } catch (Throwable t) {
+            Ln.e("### 8", t);
+            throw new AssertionError(t);
         }
     }
 
```

</details>

@FredWe @olbb @Stanislavoloshchuk @ilceramista (from https://github.com/Genymobile/scrcpy/pull/434#issuecomment-702338753) Could you test it please, and copy-paste the logs printed in the console when it is blocked. This should help to determine where it deadlocks.

For windows users, take both `scrcpy.exe` and `scrcpy-server`, and replace them in your scrcpy v1.16 release.

For other platforms, take `scrcpy-server` and [build](https://github.com/Genymobile/scrcpy/blob/master/BUILD.md#prebuilt-server) only the client.

 - [`scrcpy.exe`](https://tmp.rom1v.com/scrcpy/240/3/scrcpy.exe)
  _SHA256: 5ee3fc5f0094a50db9198ae3bb9e7fff9fd29199174f32e579e6d28bf6a69a10_
 - [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/240/3/scrcpy-server)
  _SHA256: c742ec74cf9293db9d5660e9904f26010552b6e03c21fa6f562705da130fca1d_

**Comment by rom1v** [comment]
Could you please test and paste the logs? This will help to fix the problem in the next version of scrcpy.

**Comment by rom1v** [comment]
Re-fixed, see #2656.
